### DISCUSSION TITLE
How do I make more instances of my notifiers?

### DISCUSSION THREAD

**Discussion by azwalzaman** [rrousselGit/riverpod#2594] [Q&A]
I'm currently making notifiers with this syntax

```
@riverpod
class FoodNotifier extends _$FoodNotifier {
}
```

I use this notifier to do tasks such as add, delete and update food. Each of those has their own states as well. The issue is for example when I use it to add a food item and then go back to the list of food items it would be empty. Because the state changed from `FoodGet` to `FoodAdd`. Am I approaching this wrong? If not, how do I make a different instance of this notifier?

**Comment by hasanmhallak** [comment]
I believe in this case you should use one notifier to manage one state.  
`FoodGet` probably has more than one state, specially, if you are getting the food list from API, state init, loading, failure.. etc. So is `FoodAdd`.

What I normally do is to have one repository that encapsulate all the function related to `food feature` and use as many providers as my usecases need.

**Reply by azwalzaman** [reply]
> I believe in this case you should use one notifier to manage one state. `FoodGet` probably has more than one state, specially, if you are getting the food list from API, state init, loading, failure.. etc. So is `FoodAdd`.
> 
> What I normally do is to have one repository that encapsulate all the function related to `food feature` and use as many providers as my usecases need.


How do I create more providers from the same class using this syntax?

**Reply by hasanmhallak** [reply]
you can simply use a different name for it.
```dart
@riverpod
class AddFoodNotifier extends _$AddFoodNotifier {
}

@riverpod
class GetFoodNotifier extends _$GetFoodNotifier {
}
```

The thing is you are not creating two providers from the same instance if you try to use what I suggest. You are making two separate provider for each case you have. Probably the logic for getting  food is different from adding  food.

**Reply by azwalzaman** [reply]
> I believe in this case you should use one notifier to manage one state. `FoodGet` probably has more than one state, specially, if you are getting the food list from API, state init, loading, failure.. etc. So is `FoodAdd`.
> 
> What I normally do is to have one repository that encapsulate all the function related to `food feature` and use as many providers as my usecases need.

Thanks. This was helpful. But what do I do in case I need two different instances of the same provider that do the same thing? Like one instance for a page showing foods and another for a dropdown of foods

**Reply by hasanmhallak** [reply]
hey @azwalzaman I'm not sure why you want to use two instance of the same notifier if they are doing the exact same thing. the problem with this that it would be extremely hard to keep the two of them in sync. Like you will have a page showing the food but at the same time the dropdown could still be empty as it is still getting the data.  

what I suggest is to use the same instance in order to keep everything in sync.  

if you have a provider that let's say get the food recipe based on some id, you can use the [`.family`](https://docs-v2.riverpod.dev/docs/concepts/modifiers/family) modifier.
